parameters/parser/ParamsParser.go

```diff
@@ -0,0 +1,306 @@
+	StakingCap        btcutil.Amount
+	Tag               []byte
+	CovenantPks       []*btcec.PublicKey
+	CovenantQuorum    uint32
+	UnbondingTime     uint16
+	UnbondingFee      btcutil.Amount
+	MaxStakingAmount  btcutil.Amount
+	MinStakingAmount  btcutil.Amount
+	MaxStakingTime    uint16
+	MinStakingTime    uint16
+	ConfirmationDepth uint16
+}
+
+func ParseGlobalParams(p *GlobalParams) (*ParsedGlobalParams, error) {
+	if len(p.Versions) == 0 {
+		return nil, fmt.Errorf("global params must have at least one version")
+	}
+	var parsedVersions []*ParsedVersionedGlobalParams
+
+	for _, v := range p.Versions {
+		vCopy := v
+		cv, err := parseVersionedGlobalParams(vCopy)
+
+		if err != nil {
+			return nil, fmt.Errorf("invalid params with version %d: %w", vCopy.Version, err)
+		}
+
+		// Check latest version
+		if len(parsedVersions) > 0 {
+			pv := parsedVersions[len(parsedVersions)-1]
+
+			if cv.Version != pv.Version+1 {
+				return nil, fmt.Errorf("invalid params with version %d. versions should be monotonically increasing by 1", cv.Version)
+			}
+			if cv.StakingCap < pv.StakingCap {
+				return nil, fmt.Errorf("invalid params with version %d. staking cap cannot be decreased in later versions", cv.Version)
+			}
+			if cv.ActivationHeight <= pv.ActivationHeight {
+				return nil, fmt.Errorf("invalid params with version %d. activation height cannot be overlapping between earlier and later versions", cv.Version)
+			}
+		}
+
+		parsedVersions = append(parsedVersions, cv)
+	}
+
+	return &ParsedGlobalParams{
+		Versions: parsedVersions,
+	}, nil
+}
+
+func parseVersionedGlobalParams(p *VersionedGlobalParams) (*ParsedVersionedGlobalParams, error) {
+	tag, err := hex.DecodeString(p.Tag)
+
+	if err != nil {
+		return nil, fmt.Errorf("invalid tag: %w", err)
+	}
+
+	if len(tag) != TagLen {
+		return nil, fmt.Errorf("invalid tag length, expected %d, got %d", TagLen, len(tag))
+	}
+
+	if len(p.CovenantPks) == 0 {
+		return nil, fmt.Errorf("empty covenant public keys")
+	}
+	if p.CovenantQuorum > uint64(len(p.CovenantPks)) {
+		return nil, fmt.Errorf("covenant quorum cannot be more than the amount of covenants")
+	}
+
+	quroum, err := parseUint32(p.CovenantQuorum)
+	if err != nil {
+		return nil, fmt.Errorf("invalid covenant quorum: %w", err)
+	}
+
+	var covenantKeys []*btcec.PublicKey
+	for _, covPk := range p.CovenantPks {
+		pk, err := parseCovenantPubKeyFromHex(covPk)
+		if err != nil {
+			return nil, fmt.Errorf("invalid covenant public key %s: %w", covPk, err)
+		}
+
+		covenantKeys = append(covenantKeys, pk)
+	}
+
+	maxStakingAmount, err := parseBtcValue(p.MaxStakingAmount)
+
+	if err != nil {
+		return nil, fmt.Errorf("invalid max_staking_amount: %w", err)
+	}
+
+	minStakingAmount, err := parseBtcValue(p.MinStakingAmount)
+
+	if err != nil {
+		return nil, fmt.Errorf("invalid min_staking_amount: %w", err)
+	}
+
+	if maxStakingAmount <= minStakingAmount {
+		return nil, fmt.Errorf("max-staking-amount must be larger than min-staking-amount")
+	}
```

Comment:
```suggestion
	if maxStakingAmount < minStakingAmount {
		return nil, fmt.Errorf("max-staking-amount %d must be larger than or equal to min-staking-amount %d", maxStakingAmount, minStakingAmout)
	}
```

Do we also want to allow a fixed staking amount?

Reply:
yup lets allow it 👍